File: src/main/java/javabot/model/Logs.java
package javabot.model;

import java.io.Serializable;
import java.util.Date;
import javax.persistence.Column;
import javax.persistence.Entity;
import javax.persistence.EnumType;
import javax.persistence.Enumerated;
import javax.persistence.GeneratedValue;
import javax.persistence.Id;
import javax.persistence.NamedQueries;
import javax.persistence.NamedQuery;
import javax.persistence.Table;
import javax.persistence.Temporal;
import javax.persistence.TemporalType;
import javax.persistence.Transient;

import javabot.dao.LogsDao;

@Entity
@Table(name = "logs")
@NamedQueries({
    @NamedQuery(name= LogsDao.TODAY, query = "from Logs s WHERE s.channel=:channel AND s.updated < :tomorrow AND"
        + " s.updated > :today order by updated"),
    @NamedQuery(name=LogsDao.LOGGED_CHANNELS, query="select distinct s.channel from Logs s where s.channel like '#%'")
})
public class Logs implements Serializable, Persistent {
    private Long id;
    private String nick;
    private String channel;
    private String message;
    private Date updated;

    public enum Type {
        JOIN, PART, QUIT, ACTION, KICK, BAN, MESSAGE
    }

    @Enumerated(EnumType.STRING)
    private Type type;

    @Id
    @GeneratedValue
    public Long getId() {
        return id;
    }

    public void setId(Long logsId) {
        id = logsId;
    }

    public String getNick() {
        return nick;
    }

    public void setNick(String user) {
        nick = user;
    }

    public String getChannel() {
        return channel;
    }

    public void setChannel(String chanName) {
        channel = chanName;
    }

    @Column(length = 4000)
    public String getMessage() {
        return message;
    }

    public void setMessage(String logMessage) {
        message = logMessage;
    }

    @Temporal(TemporalType.TIMESTAMP)
    public Date getUpdated() {
        return updated;
    }

    public void setUpdated(Date date) {
        updated = date;
    }

    public Type getType() {
        return type;
    }

    public void setType(Type value) {
        type = value;
    }

    @Transient
    public boolean isAction() {
        return message != null && Type.ACTION == getType();
    }

    @Transient
    public boolean isKick() {
        return message != null && Type.KICK==getType();
    }

    @Transient
    public boolean isServerMessage() {
        return message != null && Type.JOIN == getType() || Type.PART == getType() || Type.QUIT == getType();
    }
}

File: src/main/kotlin/javabot/model/Logs.kt
package javabot.model

import dev.morphia.annotations.Entity
import dev.morphia.annotations.Field
import dev.morphia.annotations.Id
import dev.morphia.annotations.Index
import dev.morphia.annotations.IndexOptions
import dev.morphia.annotations.Indexes
import dev.morphia.annotations.PrePersist
import java.time.LocalDateTime
import java.util.Locale
import javabot.model.Logs.Type
import org.bson.types.ObjectId

@Entity(value = "logs", useDiscriminator = false)
@Indexes(
    Index(
        fields = arrayOf(Field("channel"), Field("upperNick"), Field("updated")),
        options = IndexOptions(name = "seen"),
    )
)
class Logs : Persistent {
    enum class Type {
        ACTION,
        BAN,
        DISCONNECTED,
        ERROR,
        INVITE,
        JOIN,
        PART,
        KICK,
        MESSAGE,
        QUIT,
        REGISTERED,
        TOPIC,
        NICK,
    }

    @Id var id: ObjectId = ObjectId()

    var nick: String? = null

    var channel: String? = null

    lateinit var message: String

    lateinit var updated: LocalDateTime

    lateinit var type: Type

    var upperNick: String? = null

    constructor() {}

    constructor(
        nick: String?,
        message: String,
        type: Type,
        channel: String?,
        updated: LocalDateTime = LocalDateTime.now(),
    ) : this() {
        this.nick = nick
        this.message = message
        this.type = type
        this.updated = updated
        this.channel = channel
        upperNick()
    }

    fun isAction(): Boolean {
        return Type.ACTION == type
    }

    fun isKick(): Boolean {
        return Type.KICK == type
    }

    fun isServerMessage(): Boolean {
        return Type.JOIN == type || Type.PART == type || Type.QUIT == type
    }

    @PrePersist
    fun upperNick() {
        upperNick = nick?.uppercase(Locale.getDefault())
    }

    override fun toString(): String {
        return "Logs{ id=$id, type=$type, channel='$channel', updated=$updated, nick='$nick', message='$message', upperNick='$upperNick'}"
    }
}
